### PR TITLE
Improve the fit for aligned allocation.

### PR DESCRIPTION
**Pull request by interwq**
We compute the max size required to satisfy an alignment.  However this can be
quite pessimistic, especially with frequent reuse (and combined with state-based
fragmentation).  This commit adds one more fit step specific to aligned
allocations, searching in all potential fit size classes.

w/o this diff, looping ```{ ptr = mallocx(2M, align(2M); free(ptr); }``` results in no reuse of any previously freed memory:
`
0x7f917f000000
0x7f917dc00000
0x7f917d600000
0x7f917ce00000
0x7f917d200000
0x7f917c600000
0x7f917ca00000
0x7f917bc00000
0x7f917c000000
0x7f917b000000
`
with this diff:
`
0x7fa53f600000
0x7fa53f600000
0x7fa53f600000
0x7fa53f600000
0x7fa53f600000
0x7fa53f600000
0x7fa53f600000
0x7fa53f600000
0x7fa53f600000
0x7fa53f600000
`

### CODE REVIEW

**Comment by interwq** [comment]
Testing in progress. Not ready for review yet.

**Comment by interwq** [comment]
Production testing showed this diff fixed a regression with aligned allocations.

Fundamentally, two factors contributed to the issue: 1) state based fragmentation (#994) because we cannot coalesce dirty and retained extents to satisfy the alignment; and 2) in the alignment path, asking for pessimistic size which is fixed in this PR. Although testing results showed the regression being fixed with (2) alone, we probably should get down to #994 sooner.

Ready for review now.

File: src/extent.c [hunk 1]
```diff
@@ -363,6 +363,43 @@ extents_remove_locked(tsdn_t *tsdn, extents_t *extents, extent_t *extent,
...
+	for (pszind_t i = (pszind_t)bitmap_ffu(extents->bitmap,
+	    &extents_bitmap_info, (size_t)pind); i < pind_max; i =
+	    (pszind_t)bitmap_ffu(extents->bitmap, &extents_bitmap_info,
+	    (size_t)i+1)) {
+		assert(i < NPSIZES);
+		assert(!extent_heap_empty(&extents->heaps[i]));
+		extent_t *extent = extent_heap_first(&extents->heaps[i]);
+		assert(extent_size_get(extent) >= min_size);
+		uintptr_t base = (uintptr_t)extent_base_get(extent);
+		size_t esize = extent_size_get(extent);
```
Comment (davidtgoldblatt) [inline src/extent.c]: Let's call this something other than esize? We already use that in the rest of the file to mean "effective size" (i.e. size + pad).

File: src/extent.c [hunk 2]
```diff
@@ -363,6 +363,43 @@ extents_remove_locked(tsdn_t *tsdn, extents_t *extents, extent_t *extent,
...
+
+		uintptr_t next_align = ALIGNMENT_CEILING((uintptr_t)base,
```
Comment (davidtgoldblatt) [inline src/extent.c]: It looks like we don't do any overflow checking -- right?

File: src/extent.c (hunk 1 above)
Comment (interwq) [reply src/extent.c]: Good point. Will fix.

File: src/extent.c (hunk 2 above)
Comment (interwq) [reply src/extent.c]: Right. Doesn't seem to be needed (if this overflows, we won't and can't use the current extent).

Comment (davidtgoldblatt) [reply src/extent.c]: Can you add a comment with some math? I'm definitely not following.

Suppose there's a very large alignment. Then next_align will be less then base, so we'll pass the first check, so leadsize will be very big, so esize - leadsize will be very big, so we'll return an improper extent, no?

Comment (interwq) [reply src/extent.c]: Will add comments. However if the next_align is less than base, we will skip this extent in the first check, right?

Comment (interwq) [reply src/extent.c]: Hold on. That wasn’t right. Let me fix this.

**Comment by interwq** [comment]
Feedback addressed. @davidtgoldblatt you wanna take another look?

**Comment by interwq** [comment]
Only a style change was made after the last green build. Merging.

**Comment by edwinsmith** [comment]
For large size classes, I understand what went wrong and how this fixes it. For small size classes, managed in slabs with a bitmap, is it [still*] the case that sizes that are powers of 2 are going to be naturally aligned? e.g. IIUC, 4K is a small size, so it would come from a slab containing only same-sized (4K) naturally aligned objects?

* I'm saying "still" because I think with jemalloc 4 and earlier, this was the case... but correct me otherwise.

caveat: I know the APIs don't guarantee this alignment, I'm thinking about data structures that would profit from naturally aligned nodes, and what other pitfalls there might be when asking for aligned blocks.

**Comment by interwq** [comment]
For power of 2 size classes equal to or smaller than page size (default 4K), yes they will be naturally aligned of the same size, e.g 1k / 2k.

For size classes greater than page size such as 8k, no such alignment will be guaranteed (they are still small size classes, but the slab contains only 1 item and is only page aligned).

**Pull request merged**